### ISSUE TITLE
Unexpected results using ORPO trl

### ISSUE THREAD

**Issue by celsowm** [xfactlab/orpo#28]
Hi @jiwooya1000 ! Greetings from Brazil ! 🇧🇷

First of all, congrats for your amazing work ! Thanks !

Well, for studying purposes, I've created a very small dataset about a fictional city called "Auryn":

https://huggingface.co/datasets/celsowm/auryn_dpo_orpo_english

So, my goal is to "inject" new knowledge on an LLM as mistral, so I tried this:

```python
import torch, gc, sys
from peft import LoraConfig, prepare_model_for_kbit_training, PeftModel
from transformers import BitsAndBytesConfig, AutoTokenizer, AutoModelForCausalLM
from trl import setup_chat_format, ORPOConfig, ORPOTrainer
from datasets import load_dataset

torch_dtype = torch.bfloat16

if torch.cuda.get_device_capability()[0] >= 8:
    attn_implementation = "flash_attention_2" #pip install -qqq flash-attn
else:
    attn_implementation = "eager"
    
#Lora
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch_dtype,
    bnb_4bit_use_double_quant=True
#     llm_int8_enable_fp32_cpu_offload=True
)

peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=['up_proj', 'down_proj', 'gate_proj', 'k_proj', 'q_proj', 'v_proj', 'o_proj']
)

# Load Token
#model_name = "meta-llama/Meta-Llama-3-8B-Instruct"
model_name = "mistralai/Mistral-7B-v0.1"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Carregando Modelo "base"
model=AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map='auto',
    torch_dtype=torch_dtype,
    attn_implementation=attn_implementation
)

# checking params
def print_trainable_parameters(model):
    trainable_params=0
    all_params=0
    for _, param in model.named_parameters():
        all_params+=param.numel()
        if param.requires_grad:
            trainable_params+=param.numel()
    print(f"trainable params: {trainable_params} || all params: {all_params} || trainable%: {100 * trainable_params/all_params:.2f}")

print_trainable_parameters(model)

# Set chat format and feeze pretrained weights
model, tokenizer = setup_chat_format(model, tokenizer)
model = prepare_model_for_kbit_training(model)
print_trainable_parameters(model)


#load dataset
dataset_name = "celsowm/auryn_dpo_orpo_english"
dataset = load_dataset(dataset_name, split="all") #download_mode='force_redownload'

def process(row):
    
    prompt_user = {"content": row["prompt"], "role": "user"}

    row["prompt"] = tokenizer.apply_chat_template(
                    [{"role": "user", "content": row["prompt"]}],
                    tokenize=False,
                    add_generation_prompt=True
                )
    row["chosen"] = tokenizer.apply_chat_template([prompt_user, {"content": row["chosen"], "role": "assistant"}], tokenize=False)
    row["rejected"] = tokenizer.apply_chat_template([prompt_user, {"content": row["rejected"], "role": "assistant"}], tokenize=False)
    return row

dataset = dataset.map(process)
dataset = dataset.train_test_split(test_size=0.1)

orpo_args=ORPOConfig(
    learning_rate=5e-6,
    beta=0.1,
    gradient_accumulation_steps=1,
    gradient_checkpointing=True,
    lr_scheduler_type="cosine",
    max_length=1024,
    max_prompt_length=2048,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    num_train_epochs=3,
    optim="adamw_torch",
    output_dir="output/auryn_orpo_english",
    overwrite_output_dir=True,
    bf16=True,
)

trainer=ORPOTrainer(
    model=model,
    args=orpo_args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    peft_config=peft_config,
    tokenizer=tokenizer
)

trainer.train()
trainer.save_model(orpo_args.output_dir)
tokenizer.save_pretrained(orpo_args.output_dir)

#Merge
del trainer, model
gc.collect()

torch.cuda.empty_cache()

tokenizer=AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch_dtype,
    device_map="cpu",
)

model, tokenizer = setup_chat_format(model, tokenizer)

# Merge adapter with base model
model = PeftModel.from_pretrained(model, orpo_args.output_dir)
model = model.merge_and_unload()

merged_dir = "output/auryn_orpo_english_merged"
tokenizer.save_pretrained(merged_dir)
model.save_pretrained(merged_dir) 
```

After that I tried this:

```python
import torch, sys
from transformers import BitsAndBytesConfig, AutoTokenizer, AutoModelForCausalLM, pipeline

torch_dtype = torch.bfloat16

if torch.cuda.get_device_capability()[0] >= 8:
    attn_implementation = "flash_attention_2" #pip install -qqq flash-attn
else:
    attn_implementation = "eager"
    
#Quantização com QLora
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch_dtype,
    bnb_4bit_use_double_quant=True,
    llm_int8_enable_fp32_cpu_offload=True
)

model_name = "output/auryn_orpo_english_merged"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Carregando Modelo "base"
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map='auto',
    torch_dtype=torch_dtype,
    attn_implementation=attn_implementation
)

# Create a conversation pipeline
conversation = pipeline("conversational", model=model, tokenizer=tokenizer)

# Define the chat history
chat_history = [
    {"role": "user", "content": "Who founded the city of Auryn?"},
]

# Generate a response
response = conversation(chat_history)
print(response)
```

The response:

```
Conversation id: b82f1b79-6c21-4f6e-8af5-1ec587e21ef1
user: Who founded the city of Auryn?
assistant: user
Who founded the city of Auryn?
```

I was expecting something like: Jonathan Auryn (chosen)

So, what did I do wrong?
Can you help me?

Thanks in advance !

**Comment by nlee-208** [comment]
Hey @celsowm, it's hard to exactly trace why the training didn't go well but I'll suggest a few possible solutions and hope these work :)

- Try updating to the latext trl dev version as we have some fixing of the implementations
- Check whether the chat template is properly utilized as the Mistral-v0.1 wouldn't have a chat template itself (it'll probably use the basic chat template). Make sure the trained and evaluated models are using the same chat template as chat models are supersensitive to these
- Increased learning rate such as 5e-4 or 5e-5 as you're using LoRA + the dataset size is relatively small
- Increased beta value to ensure the margin of the correct and incorrect knowledges apart 

Thank you for the interest and hope these work..!

**Comment by celsowm** [comment]
First of all, thanks for reply @nlee-208 

> Hey @celsowm, it's hard to exactly trace why the training didn't go well but I'll suggest a few possible solutions and hope these work :)
> 
>     * Try updating to the latext trl dev version as we have some fixing of the implementations
pip install -U trl and no updates in my case
>     * Check whether the chat template is properly utilized as the Mistral-v0.1 wouldn't have a chat template itself (it'll probably use the basic chat template). Make sure the trained and evaluated models are using the same chat template as chat models are supersensitive to these
okay, Changed to mistralai/Mistral-7B-Instruct-v0.2
>     * Increased learning rate such as 5e-4 or 5e-5 as you're using LoRA + the dataset size is relatively small
Was already 5e-6
>     * Increased beta value to ensure the margin of the correct and incorrect knowledges apart
okay, Increased to 0.4
> 
> Thank you for the interest and hope these work..!

Unfortunately, even with all those changes, the result still the same :(

**Comment by jiwooya1000** [comment]
Hello @celsowm! Thank you for the detailed reply.

I was looking through your code above and found one thing that needs to be changed.
But before that, we recommend you use the latest `dev` version of trl, which needs to be installed from git right away!

For TRL, the dataset format should be (`prompt`, `chosen response`, `rejected response`).
However, the dataset preprocessing function in your code seems to make the `chosen response` field by concatenating the prompt and the chosen response (which is the same for the `rejected response` field).

Therefore, I think you could check the formatting in the response field!

```
def process(row):
    
    prompt_user = {"content": row["prompt"], "role": "user"}

    row["prompt"] = tokenizer.apply_chat_template(
                    [{"role": "user", "content": row["prompt"]}],
                    tokenize=False,
                    add_generation_prompt=True
                )
    row["chosen"] = tokenizer.apply_chat_template([prompt_user, {"content": row["chosen"], "role": "assistant"}], tokenize=False)
    row["rejected"] = tokenizer.apply_chat_template([prompt_user, {"content": row["rejected"], "role": "assistant"}], tokenize=False)
    return row
```

**Comment by celsowm** [comment]
> Hello @celsowm! Thank you for the detailed reply.
> 
> I was looking through your code above and found one thing that needs to be changed. But before that, we recommend you use the latest `dev` version of trl, which needs to be installed from git right away!
> 
> For TRL, the dataset format should be (`prompt`, `chosen response`, `rejected response`). However, the dataset preprocessing function in your code seems to make the `chosen response` field by concatenating the prompt and the chosen response (which is the same for the `rejected response` field).
> 
> Therefore, I think you could check the formatting in the response field!
> 
> ```
> def process(row):
>     
>     prompt_user = {"content": row["prompt"], "role": "user"}
> 
>     row["prompt"] = tokenizer.apply_chat_template(
>                     [{"role": "user", "content": row["prompt"]}],
>                     tokenize=False,
>                     add_generation_prompt=True
>                 )
>     row["chosen"] = tokenizer.apply_chat_template([prompt_user, {"content": row["chosen"], "role": "assistant"}], tokenize=False)
>     row["rejected"] = tokenizer.apply_chat_template([prompt_user, {"content": row["rejected"], "role": "assistant"}], tokenize=False)
>     return row
> ```

Hi @jiwooya1000 and @nlee-208 !

I've update to trl-0.8.7.dev0 and also remove the concatenation and....
IT WORKED !!!! YEAHH !!!

I thought the concatenation in my case would be mandatory because, in my dataset, I've just used strings:

![image](https://github.com/xfactlab/orpo/assets/369336/38f4a8f1-2bdc-40b5-99ff-cfc2a4a809bc)

And mlabonne used array of dicts including the prompt on chosen and reject:

![image](https://github.com/xfactlab/orpo/assets/369336/79d79814-faf0-4dec-8493-f7fd15133c93)

But, anyway ! It worked perfect now !

```
Conversation id: 375494ff-a41c-439d-be22-20b5a4fd452e
user: Who founded the city of Auryn?
assistant: assistant
Jonathan Auryn
```

So the final code was:

```
import torch, gc, sys
from peft import LoraConfig, prepare_model_for_kbit_training, PeftModel
from transformers import BitsAndBytesConfig, AutoTokenizer, AutoModelForCausalLM
from trl import setup_chat_format, ORPOConfig, ORPOTrainer
from datasets import load_dataset

torch_dtype = torch.bfloat16

if torch.cuda.get_device_capability()[0] >= 8:
    attn_implementation = "flash_attention_2" #pip install -qqq flash-attn
else:
    attn_implementation = "eager"
    
    
#Lora
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type="nf4",
    bnb_4bit_compute_dtype=torch_dtype,
    bnb_4bit_use_double_quant=True
#     llm_int8_enable_fp32_cpu_offload=True
)

peft_config = LoraConfig(
    r=16,
    lora_alpha=32,
    lora_dropout=0.05,
    bias="none",
    task_type="CAUSAL_LM",
    target_modules=['up_proj', 'down_proj', 'gate_proj', 'k_proj', 'q_proj', 'v_proj', 'o_proj']
)

# Load Token
#model_name = "meta-llama/Meta-Llama-3-8B-Instruct"
model_name = "mistralai/Mistral-7B-Instruct-v0.2"
tokenizer = AutoTokenizer.from_pretrained(model_name)

# Carregando Modelo "base"
model=AutoModelForCausalLM.from_pretrained(
    model_name,
    quantization_config=bnb_config,
    device_map='auto',
    torch_dtype=torch_dtype,
    attn_implementation=attn_implementation
)

# checking params
def print_trainable_parameters(model):
    trainable_params=0
    all_params=0
    for _, param in model.named_parameters():
        all_params+=param.numel()
        if param.requires_grad:
            trainable_params+=param.numel()
    print(f"trainable params: {trainable_params} || all params: {all_params} || trainable%: {100 * trainable_params/all_params:.2f}")

print_trainable_parameters(model)

# Set chat format and feeze pretrained weights
model, tokenizer = setup_chat_format(model, tokenizer)
model = prepare_model_for_kbit_training(model)
print_trainable_parameters(model)

#load dataset
dataset_name = "celsowm/auryn_dpo_orpo_english"
dataset = load_dataset(dataset_name, split="all") #download_mode='force_redownload'

def process(row):

    row["prompt"] = tokenizer.apply_chat_template(
                    [{"role": "user", "content": row["prompt"]}],
                    tokenize=False,
                    add_generation_prompt=True
                )
    row["chosen"] = tokenizer.apply_chat_template([{"content": row["chosen"], "role": "assistant"}], tokenize=False)
    row["rejected"] = tokenizer.apply_chat_template([{"content": row["rejected"], "role": "assistant"}], tokenize=False)
    return row

dataset = dataset.map(process)
dataset = dataset.train_test_split(test_size=0.1)

orpo_args=ORPOConfig(
    learning_rate=5e-6,
    beta=0.4,
    gradient_accumulation_steps=1,
    gradient_checkpointing=True,
    lr_scheduler_type="cosine",
    max_length=1024,
    max_prompt_length=2048,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    num_train_epochs=3,
    optim="adamw_torch",
    output_dir="output/auryn_orpo_english",
    overwrite_output_dir=True,
    bf16=True,
)

trainer=ORPOTrainer(
    model=model,
    args=orpo_args,
    train_dataset=dataset["train"],
    eval_dataset=dataset["test"],
    peft_config=peft_config,
    tokenizer=tokenizer
)

trainer.train()
trainer.save_model(orpo_args.output_dir)
tokenizer.save_pretrained(orpo_args.output_dir)

#Merge
del trainer, model
gc.collect()

torch.cuda.empty_cache()

tokenizer=AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(
    model_name,
    low_cpu_mem_usage=True,
    return_dict=True,
    torch_dtype=torch_dtype,
    device_map="cpu",
)

model, tokenizer = setup_chat_format(model, tokenizer)

# Merge adapter with base model
model = PeftModel.from_pretrained(model, orpo_args.output_dir)
model = model.merge_and_unload()

merged_dir = "output/auryn_orpo_english_merged"
tokenizer.save_pretrained(merged_dir)
model.save_pretrained(merged_dir)
```

 Thanks a lot
